### PR TITLE
Add NIH account linking

### PR DESCRIPTION
**Pull request by breilly2**
Resolves: #1372

Testing in dev uses a mock-shibboleth. You can give it any name and it will apply the callback with a token for that name.

To reset your linked status, find your Google subject ID, connect to the thurloe mysql database:
```
docker run -v $HOME:/root --rm -it broadinstitute/dsde-toolbox:dev mysql-connect.sh -p firecloud -e dev -a thurloe
```
and delete some rows:
```
delete from KEY_VALUE_PAIR where USER_ID = '<subject ID>' and `KEY` IN ('linkedNihUsername', 'linkExpireTime');
```

### CODE REVIEW

File: src/libs/ajax.js [hunk 1]
```diff
@@ -238,6 +238,24 @@ const User = signal => ({
 
   postNpsResponse: async body => {
     return fetchRex('npsResponses/create', _.mergeAll([authOpts(), jsonBody(body), { signal, method: 'POST' }]))
+  },
+
+  getNihStatus: async () => {
+    try {
+      const res = await fetchOrchestration('api/nih/status', _.merge(authOpts(), { signal }))
+      return res.json()
+    } catch (error) {
```
Comment (breilly2) [inline src/libs/ajax.js]: Is there a better way to express this special handling of 404? This feels a bit wordy to me.

File: src/libs/utils.js [hunk 2]
```diff
@@ -259,7 +259,11 @@ export const withBusyState = _.curry((setBusy, fn) => async (...args) => {
   }
 })
 
-// Performs the given effect, but only on component mount
+/**
```
Comment (breilly2) [inline src/libs/utils.js:262]: Drive-by documentation tweaking. This is my understanding based on discussion with Isaac. Please offer corrections/suggestions if you have them.

File: src/pages/Profile.js [hunk 3]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
+  const linkNihAccount = async nihToken => {
+    setLinking(true)
+    const status = await Ajax().User.linkNihAccount(nihToken)
+    setNihStatus(status)
+    setLinking(false)
+  }
+
+
+  /*
+   * Hooks
```
Comment (breilly2) [inline src/pages/Profile.js:66]: This was fun!

File: src/pages/Profile.js [hunk 4]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
 
+const NihLink = ({ nihToken }) => {
+  const loadNihStatus = async () => {
+    const status = await Ajax().User.getNihStatus()
+    setNihStatus(status)
+  }
+
+  const linkNihAccount = async nihToken => {
+    setLinking(true)
+    const status = await Ajax().User.linkNihAccount(nihToken)
```
Comment (breilly2) [inline src/pages/Profile.js]: Ah, I should use `Utils.withBusyState` here.

File: src/libs/ajax.js (hunk 1 above)
Comment (elizzhao) [reply src/libs/ajax.js]: My impression is we usually don't do the error handling directly in the ajax call. Perhaps you could pull the try catch into where you call it in profile in `loadNihStatus()`, and then directly do `setNihStatus({})` if you get a 404 error?

see my other comment for where I'm talking about in case I wasn't being clear here

File: src/pages/Profile.js [hunk 5]
```diff
@@ -58,6 +59,107 @@ const styles = {
 }
 
 
+const NihLink = ({ nihToken }) => {
+  const loadNihStatus = async () => {
+    const status = await Ajax().User.getNihStatus()
```
Comment (elizzhao) [inline src/pages/Profile.js]: You could put the error handling here--maybe using `withErrorReporting`?

File: src/pages/Profile.js [hunk 6]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
+    setNihStatus(status)
+  }
+
+  const linkNihAccount = async nihToken => {
```
Comment (elizzhao) [inline src/pages/Profile.js]: should we have a error handling here since we're making an ajax call?

File: src/pages/Profile.js [hunk 7]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
+    !!linkedNihUsername && div({ style: { display: 'flex', flexDirection: 'column', width: '35rem' } }, [
+      div({ style: { display: 'flex' } }, [
+        div({ style: { flex: 1 } }, ['Username:']),
+        div({ style: { flex: 2 } }, [linkedNihUsername])
+      ]),
+      div({ style: { display: 'flex' } }, [
+        div({ style: { flex: 1 } }, ['Link Expiration:']),
+        div({ style: { flex: 2 } }, [
+          Utils.makeCompleteDate(linkExpireTime * 1000),
+          makeAccountLinkLink('Log-In to NIH to re-link your account')
```
Comment (elizzhao) [inline src/pages/Profile.js]: Should we still show this if their link hasn't expired yet?

File: src/pages/Profile.js [hunk 8]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
+
+
+  /*
+   * Render
+   */
+  return h(Fragment, [
+    div({ style: styles.form.title }, [
+      'NIH Account',
+      h(InfoBox, { style: { marginLeft: '0.5rem' } }, [
+        'Linking with eRA Commons will allow FireCloud to automatically determine if you can access controlled datasets hosted in FireCloud (ex. TCGA) based on your valid dbGaP applications.'
```
Comment (katevoss) [inline src/pages/Profile.js]: Terra, not FireCloud 
```suggestion
        'Linking with eRA Commons will allow Terra to automatically determine if you can access controlled datasets hosted in Terra (ex. TCGA) based on your valid dbGaP applications.'
```

File: src/libs/ajax.js (hunk 1 above)
Comment (breilly2) [reply src/libs/ajax.js]: You're absolutely right. I keep trying to do things like this here when it should be up to the caller.

File: src/pages/Profile.js (hunk 7 above)
Comment (breilly2) [reply src/pages/Profile.js]: This is what FireCloud does. Note that it says "re-link". I'm guessing there's a good reason why someone might need to do this.

Comment (dmohs) [reply src/pages/Profile.js]: Just as likely a historical artifact.

One thing that FireCloud does not have is the ability to unlink—Example, I no longer need the link but I see the "expired" banner perpetually. A worthy follow-on story, in my opinion.

Comment (breilly2) [reply src/pages/Profile.js]: Agree. I didn't notice any existing unlink API, otherwise I'd add it right now.

Comment (katevoss) [inline src/pages/Profile.js]:
```suggestion
          makeAccountLinkLink('Log in to NIH to re-link your account')
```

File: src/pages/Profile.js [hunk 9]
```diff
@@ -58,6 +59,107 @@ const styles = {
...
+          Utils.makeCompleteDate(linkExpireTime * 1000),
+          makeAccountLinkLink('Log-In to NIH to re-link your account')
+        ])
+      ]),
+      _.flow(
+        _.sortBy('name'),
+        _.map(makeDatasetAuthStatus)
+      )(datasetPermissions)
+    ]),
+    !linkedNihUsername && makeAccountLinkLink('Log-In to NIH to link your account')
```
Comment (katevoss) [inline src/pages/Profile.js]:
```suggestion
    !linkedNihUsername && makeAccountLinkLink('Log in to NIH to link your account')
```

File: src/pages/Profile.js [hunk 10]
```diff
@@ -58,6 +60,120 @@ const styles = {
...
+    } else {
+      loadNihStatus()
+    }
+  })
+
+
+  /*
+   * Render helpers
+   */
+  const makeAccountLinkLink = label => {
```
Comment (katevoss) [inline src/pages/Profile.js]: Is the "Link" intentionally repeated, `LinkLink`?

**Review by katevoss — COMMENTED** [review]
Looks good, I have a couple of comments and questions.

File: src/pages/Profile.js (hunk 10 above)
Comment (breilly2) [reply src/pages/Profile.js]: :) Yes, it's a function that makes a link for account linking. I agree it does look like it could be a mistake. I'll try to think of a better name. Suggestions welcome.

Comment (katevoss) [reply src/pages/Profile.js]: Hmm what about `makeLinkToAccountLink` or `makeLinkToAccount`?

File: src/pages/Profile.js [hunk 11]
```diff
@@ -58,6 +60,120 @@ const styles = {
...
+      } catch (error) {
+        if (error.status === 404) setNihStatus({})
+        else throw error
+      }
+    })
+
+    if (nihToken) {
+      // Clear the query string, but use replace so the back button doesn't take the user back to the token
+      Nav.history.replace({ search: '' })
+      linkNihAccount(nihToken)
```
Comment (panentheos) [inline src/pages/Profile.js:98]: TOL: Rather than doing this on the profile page, I wonder if it would be cleaner to have a separate 'page' that just performed the linking and then redirected to the profile.

File: src/pages/Profile.js [hunk 12]
```diff
@@ -58,6 +60,120 @@ const styles = {
...
+
+
+  /*
+   * Render helpers
+   */
+  const makeAccountLinkLink = label => {
+    const nihRedirectUrl = `${window.location.protocol}//${window.location.host}/#profile?nih-username-token={token}`
+
+    return link({
+      href: `${getConfig().shibbolethUrlRoot}/link-nih-account?redirect-url=${encodeURIComponent(nihRedirectUrl)}`,
```
Comment (panentheos) [inline src/pages/Profile.js]: We should use the `qs` library to generate query strings, since it will automatically handle url encoding and combining multiple parameters (if needed)

File: src/pages/Profile.js [hunk 13]
```diff
@@ -58,6 +60,120 @@ const styles = {
...
+      loadNihStatus()
+    }
+  })
+
+
+  /*
+   * Render helpers
+   */
+  const makeAccountLinkLink = label => {
+    const nihRedirectUrl = `${window.location.protocol}//${window.location.host}/#profile?nih-username-token={token}`
```
Comment (panentheos) [inline src/pages/Profile.js]: `window.location.origin` should include both the protocol and the host. Also, for the path portion it would be preferable to use `Nav.getLink` since this is one of our own URLs.

File: src/pages/Profile.js (hunk 10 above)
Comment (breilly2) [reply src/pages/Profile.js]: Going with `makeLinkForAccountLinking`. Work for you?

File: src/pages/Profile.js (hunk 11 above)
Comment (breilly2) [reply src/pages/Profile.js:98]: Possibly. There is some trickiness around doing the redirect regardless of outcome, reporting the error if there is one, and capturing the result of the link call which avoids calling the status endpoint immediately after linking. That last point probably isn't a huge deal, but I'm looking at some refactoring to put the NIH link status in `authStore` so that the global link check and the rendering on the profile page can work from the same data. It might be easier to act on this suggestion after that is done. Does that work for you @panentheos?

Comment (panentheos) [reply src/pages/Profile.js:98]: Sounds good!

**Review by elizzhao — APPROVED** [review]
Looks good to me!

File: src/pages/Profile.js (hunk 10 above)
Comment (breilly2) [reply src/pages/Profile.js]: @katevoss FYI, I'm probably renaming this again soon... probably when adding the warning banner. I'm thinking `renderShibbolethLink` because that's what it is. Someone reading the code already knows it's for account linking because of the label passed to it, e.g. "Log in to NIH to link your account".
